src/forms/Typeahead.jsx

```diff
@@ -14,32 +14,82 @@ const defaultItemToString = itemValue => (typeof itemValue === 'string' ? itemVa
  * @module Typeahead
  */
 class Typeahead extends React.PureComponent {
+	constructor(props) {
+		super(props);
+
+		this.handleSingleSelection = this.handleSingleSelection.bind(this);
+		this.handleMultiselectSelect = this.handleMultiselectSelect.bind(this);
+		this.stateReducer = this.stateReducer.bind(this);
+	}
+
+	handleMultiselectSelect(item) {
+		this.props.onSelect(item, [...this.props.multiSelectValues, item]);
+	}
+
+	handleSingleSelection(selectedItem, stateAndHelpers) {
+		!this.props.multiSelect && this.props.onSelect && this.props.onSelect(selectedItem, stateAndHelpers);
+	}
+
+	stateReducer(state, changes) {
+		switch (changes.type) {
+			case Downshift.stateChangeTypes.keyDownEnter:
+			case Downshift.stateChangeTypes.clickItem:
+				this.handleMultiselectSelect(changes.selectedItem);
+
+				return {
+					...changes,
+					selectedItem: this.props.multiSelectValues,
+					isOpen: this.props.openOnFocus,
+				};
+
+			case Downshift.stateChangeTypes.blurInput:
+				return {
+					...changes,
+					isOpen: this.props.openOnFocus,
+				};
+
+			default:
+				return changes;
+		}
+	}
 
 	render() {
 		const {
 			inputProps,
 			height,
 			items,
+			multiSelect,
+			multiSelectValues,
+			openOnFocus,
 			itemToString,
 			...other
 		} = this.props;
 
+		delete other.onSelect;
+
 		return (
 			<Downshift
+				onSelect={this.handleSingleSelection}
+				selectedItem={multiSelectValues}
+				stateReducer={multiSelect && this.stateReducer}
```

Comment:
This is incorrect - I only need to use `stateReducer` when we want to keep the Typeahead menu open after selection